File: packages/core/src/rules/inline.ts
import type { Text } from 'domhandler';
import type { Rule } from '../types.js';

/**
 * Built-in rules for inline HTML elements.
 *
 * Handles bold (`strong`, `b`), italic (`em`, `i`), strikethrough (`del`, `s`),
 * inline code, links, images, abbreviations, subscript, and superscript.
 * Delimiter styles for bold and italic are configurable via the `strongDelimiter`
 * and `emDelimiter` options.
 */
export const inlineRules: Rule[] = [
    {
        filter: ['strong', 'b'],
        replacement: ({ convertChildren, node, options }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            return `${options.strongDelimiter}${content}${options.strongDelimiter}`;
        },
        priority: 0
    },

    {
        filter: ['em', 'i'],
        replacement: ({ convertChildren, node, options }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            return `${options.emDelimiter}${content}${options.emDelimiter}`;
        },
        priority: 0
    },

    {
        filter: ['del', 's', 'strike'],
        replacement: ({ convertChildren, node }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            return `~~${content}~~`;
        },
        priority: 0
    },

    {
        filter: 'code',
        replacement: ({ node, insidePre }) => {
            if (insidePre) return; // let <pre> rule handle it
            const text = node.children.map(c => ('data' in c ? (c as Text).data : '')).join('');
            if (!text) return null;
            // Use double backticks if content contains a backtick
            const delimiter = text.includes('`') ? '``' : '`';
            const padded = text.startsWith('`') || text.endsWith('`') ? ` ${text} ` : text;
            return `${delimiter}${padded}${delimiter}`;
        },
        priority: 0
    },

    {
        filter: 'a',
        replacement: ({ node, convertChildren, options }) => {
            const href = node.attribs.href || '';
            const content = convertChildren(node).trim();
            if (!content) return null;

            const resolvedHref =
                options.baseUrl && href && !href.startsWith('http') && !href.startsWith('#') ? new URL(href, options.baseUrl).href : href;

            const title = node.attribs.title;
            if (title) {
                return `[${content}](${resolvedHref} "${title}")`;
            }
            return `[${content}](${resolvedHref})`;
        },
        priority: 0
    },

    {
        filter: 'img',
        replacement: ({ node, options }) => {
            const src = node.attribs.src || '';
            const alt = node.attribs.alt || '';
            const title = node.attribs.title;

            const resolvedSrc =
                options.baseUrl && src && !src.startsWith('http') && !src.startsWith('data:') ? new URL(src, options.baseUrl).href : src;

            if (title) {
                return `![${alt}](${resolvedSrc} "${title}")`;
            }
            return `![${alt}](${resolvedSrc})`;
        },
        priority: 0
    },

    {
        filter: ['abbr', 'mark'],
        replacement: ({ convertChildren, node }) => {
            return convertChildren(node);
        },
        priority: 0
    },

    {
        filter: 'sub',
        replacement: ({ convertChildren, node }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            return `~${content}~`;
        },
        priority: 0
    },

    {
        filter: 'sup',
        replacement: ({ convertChildren, node }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            return `^${content}^`;
        },
        priority: 0
    }
];

File: python/src/markdown_for_agents/rules/inline.py
from __future__ import annotations

from urllib.parse import urljoin

from .._types import Rule, RuleContext
from ..dom.nodes import Text

inline_rules: list[Rule] = [
    Rule(
        filter=["strong", "b"],
        replacement=lambda ctx: _strong_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter=["em", "i"],
        replacement=lambda ctx: _em_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter=["del", "s", "strike"],
        replacement=lambda ctx: _del_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="code",
        replacement=lambda ctx: _code_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="a",
        replacement=lambda ctx: _a_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="img",
        replacement=lambda ctx: _img_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter=["abbr", "mark"],
        replacement=lambda ctx: ctx.convert_children(ctx.node),
        priority=0,
    ),
    Rule(
        filter="sub",
        replacement=lambda ctx: _sub_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="sup",
        replacement=lambda ctx: _sup_replacement(ctx),
        priority=0,
    ),
]


def _strong_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    d = ctx.options.strong_delimiter
    return f"{d}{content}{d}"


def _em_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    d = ctx.options.em_delimiter
    return f"{d}{content}{d}"


def _del_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    return f"~~{content}~~"


def _code_replacement(ctx: RuleContext) -> str | None:
    from .._types import SKIP as _SKIP

    if ctx.inside_pre:
        return _SKIP  # type: ignore[return-value]
    text = "".join(c.data if isinstance(c, Text) else "" for c in ctx.node.children)
    if not text:
        return None
    # Use double backticks if content contains a backtick
    delimiter = "``" if "`" in text else "`"
    padded = f" {text} " if text.startswith("`") or text.endswith("`") else text
    return f"{delimiter}{padded}{delimiter}"


def _a_replacement(ctx: RuleContext) -> str | None:
    href = ctx.node.attribs.get("href", "")
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None

    resolved_href = href
    if ctx.options.base_url and href and not href.startswith("http") and not href.startswith("#"):
        resolved_href = urljoin(ctx.options.base_url, href)

    title = ctx.node.attribs.get("title")
    if title:
        return f'[{content}]({resolved_href} "{title}")'
    return f"[{content}]({resolved_href})"


def _img_replacement(ctx: RuleContext) -> str:
    src = ctx.node.attribs.get("src", "")
    alt = ctx.node.attribs.get("alt", "")
    title = ctx.node.attribs.get("title")

    resolved_src = src
    if ctx.options.base_url and src and not src.startswith("http") and not src.startswith("data:"):
        resolved_src = urljoin(ctx.options.base_url, src)

    if title:
        return f'![{alt}]({resolved_src} "{title}")'
    return f"![{alt}]({resolved_src})"


def _sub_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    return f"~{content}~"


def _sup_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    return f"^{content}^"
